kubernetes/mutating-admission/main.rego

```diff
@@ -0,0 +1,179 @@
+package library.kubernetes.admission.mutating
+
+###########################################################################
+# Implementation of the k8s admission control external webhook interface,
+# combining validating and mutating admission controllers
+###########################################################################
+
+main = {
+	"apiVersion": "admission.k8s.io/v1beta1",
+	"kind": "AdmissionReview",
+	"response": response,
+}
+
+default response = {"allowed": true}
+
+# non-patch response i.e. validation response
+response = x {
+	count(patch) = 0
+
+	x := {
+		"allowed": false,
+		"status": {"reason": reason},
+	}
+
+	reason = concat(", ", deny)
+	reason != ""
+}
+
+# patch response i.e. mutating respone
+response = x {
+	count(patch) > 0
+
+	# if there are missing leaves e.g. trying to add a label to something that doesn't
+	# yet have any, we need to create the leaf nodes as well
+
+	fullPatches := ensureParentPathsExist(cast_array(patch))
+
+	x := {
+		"allowed": true,
+		"patchType": "JSONPatch",
+		"patch": base64url.encode(json.marshal(fullPatches)),
+	}
+}
+
+isValidRequest {
+	# not sure if this might be a race condition, it might get called before 
+	# all the validation rules have been run
+	count(deny) = 0
+}
+
+isCreateOrUpdate {
+	isCreate
+}
+
+isCreateOrUpdate {
+	isUpdate
+}
+
+isCreate {
+	input.request.operation == "CREATE"
+}
+
+isUpdate {
+	input.request.operation == "UPDATE"
+}
+
+###########################################################################
+# PATCH helpers 
+# Note: These rules assume that the input is an object
+# not an AdmissionRequest, because labels and annotations 
+# can apply to various sub-objects within a request
+# So from the context of an AdmissionRequest they need to
+# be called like
+#   hasLabelValue[["foo", "bar"]] with input as input.request.object
```

Comment:
I'm wondering if these helpers are necessary. The same line could be written as:

```
input.request.object.metadata.labels["foo"] = "bar"
```

If we want to keep these, it would be useful to explain why they're included.

Reply:
They were really just for readability, as I've been showing people who are new to rego and it was easy to talk to. Not married to them, but manipulating labels and annotations will be a pretty common use case.

Also, given that labels and annotations can appear on lower level objects it just keeps the expressions slightly shorter and clearer. Like if it were on a Pod template in a Deployment it would be 

```
podTemplate := input.request.object.spec.template
hasLabelValue["foo", "bar"] with input as podTemplate
```

rather than 

```
input.request.object.spec.template.metadata.labels["foo"] = "bar"
```
Your call. Is there much computational overhead for using variables and docs/functions for things like this over just single statements? If it adds compute overhead then we probably shouldn't encourage it.